### PR TITLE
[feature] add calculate-per-token-loss for eb

### PR DESCRIPTION
**Pull request by Wennie396**
### PR Category
Execute Infrastructure

### PR Types
New features

### Description
增加calculate-per-token-loss开关

### 是否引起精度变化
否

### CODE REVIEW

File: src/paddlefleet/context_parallel_utils.py [hunk 1]
```diff
@@ -552,10 +583,22 @@ def forward(ctx, input_tensor, axis=0, mode="dualchunk_allgather"):
...
-        return all_gather_balance(grad_output, axis=ctx.axis, group=ctx.group)
+        else:
+            grad = all_gather_balance(grad_output, axis=ctx.axis, group=ctx.group)
+        # Per-token loss + this specific call site opted in (embedding output): the
+        # all-gather put cp identical full-sequence gradient copies on every CP
+        # rank, so the upstream embedding parameter would be counted cp times under
+        # the pure-SUM reduce. Divide them back out here; ``ctx.group`` carries this
+        # microbatch's own CP size. Both conditions are required, so every other
+        # call site and the legacy path are untouched. See loss_grad_normalization.md 3.5/6.1.
+        if getattr(ctx, "scale_grad_by_cp", False) and _per_token_grad_scale_enabled():
```
Comment (risemeup1111) [inline src/paddlefleet/context_parallel_utils.py:597]: P1

新增的除法分支在本仓库内不可达：`set_per_token_grad_scale()`（本文件 76-79 行）除定义外没有任何调用点，`_PER_TOKEN_GRAD_SCALE` 恒为 `False`，因此 `grad / nranks` 永远不会执行。上方注释写明该开关应由 trainer/model init 按 `config.calculate_per_token_loss` 设置，但本 PR 未提供该设置点；后果是用户一旦打开 `calculate_per_token_loss`，按本处注释自身的分析，embed_tokens 的梯度仍会被重复计入 cp 次，per-token 训练得到的 embedding 梯度静默偏大。请在本仓库内补上初始化调用（例如模型/trainer init 依据 config 调用 `set_per_token_grad_scale(...)`），或明确该 setter 由外部调用方负责并在文档中说明契约。

File: src/paddlefleet/models/common/language_loss/language_loss.py [hunk 2]
```diff
@@ -316,7 +346,11 @@ def forward_impl(self, logits: Tensor | tuple, labels: Tensor) -> Tensor:
             loss = paddle.sum(
                 loss.cast(paddle.float32).reshape([-1]) * lossmask
             )
-            loss = loss / lossmask.sum()
+            if not getattr(self.config, "calculate_per_token_loss", False):
```
Comment (risemeup1111) [inline src/paddlefleet/models/common/language_loss/language_loss.py:349]: P1

开关打开后这里返回的是本 rank 分片的原始 token 求和，但全局 token 分母在本仓库内没有任何实现：新增的 `pop_per_token_local_count()`（55-66 行）同样没有调用点，`_PER_TOKEN_LOCAL_COUNT` 只累加、从不读取或清零，注释所说的“由 trainer 在 step 末尾归一化”在本 PR 中没有对应代码。`calculate_per_token_loss` 是 `transformer_config.py` 中既有的公开字段，任何用户仅打开它就会让 loss/梯度尺度变化约等于本 rank 的非 padding token 数，且没有任何补偿或告警；该计数还会在每个 MTP depth（`forward` 逐 depth 调用 `_forward`）与每个 micro-batch（含 eval）累加，与“全局非 padding token 数”的语义并不一致。建议在本 PR 内补上消费该计数的归一化逻辑，或在缺少训练侧配套代码时显式报错并在文档中标注该限制。

File: src/paddlefleet/context_parallel_utils.py [hunk 3]
```diff
@@ -532,9 +551,21 @@ class ContextParallelScatterOp(PyLayer):
     """
 
     @staticmethod
-    def forward(ctx, input_tensor, axis=0, mode="dualchunk_allgather"):
+    def forward(ctx, input_tensor, axis=0, mode="dualchunk_allgather", scale_grad_by_cp=False):
```
Comment (risemeup1111) [inline src/paddlefleet/context_parallel_utils.py:554]: P2

本 PR 的改动未通过仓库的 ruff 检查，当前 head 的 Codestyle Check 因此失败：`ruff check` 会修复 1 处（`language_loss.py` 新增注释块前多出的空行），`ruff format` 会重排本 PR 改动的全部 4 个文件——本行函数签名需要按参数逐行展开，`context_parallel_utils.py:590/597`、`language_loss.py:408`、`gpt_embedding.py:695-704/774`、`multi_token_prediction.py:1484` 同样需要换行。请在本地执行 `pre-commit run --all-files`（仓库 line-length=80）后重新推送，否则该必过检查会持续为红。

File: src/paddlefleet/models/common/language_loss/language_loss.py [hunk 4]
```diff
@@ -441,7 +481,15 @@ def _cast_loss_func(logits, labels):
             # EC-compat: line-wise loss (per-sample mean then average across samples)
             # EC's ErniemmPretrainingCriterion recomputes loss as line-wise when task_id
             # is present, which changes the value due to division by (count + 1e-6).
-            if self.config.gpt_model_use_experimental_version:
+            if getattr(self.config, "calculate_per_token_loss", False):
```
Comment (risemeup1111) [inline src/paddlefleet/models/common/language_loss/language_loss.py:484]: P2

该开关没有任何测试覆盖：仓库内没有任何测试引用 `calculate_per_token_loss`，本处与 `forward_impl` 的 per-token 分支、以及 `ContextParallelScatterOp` 的 `scale_grad_by_cp` 分支都没有对应用例，而这条路径同时改变 loss 归一化与 embedding 梯度尺度，属于“默认关闭、打开即错却无人发现”的风险点。建议在 `tests/multi_card_tests` 中（仓库已有 CP scatter 相关用例可参考）补一个开/关开关的对照用例，至少校验 CP>1 时 loss 尺度与 embed_tokens 梯度的比例关系。

File: src/paddlefleet/context_parallel_utils.py [hunk 5]
```diff
@@ -64,6 +64,25 @@ def uses_cutedsl_backend(*args, **kwargs):
         ) from exc
 
 
+# calculate_per_token_loss switch, read by ContextParallelScatterOp.backward to
+# decide whether to divide the all-gathered gradient by the CP group size (see the
+# comment there and loss_grad_normalization.md 3.5/6.1). Module-level rather than a
```
Comment (risemeup1111) [inline src/paddlefleet/context_parallel_utils.py:69]: P3

注释多处引用 `loss_grad_normalization.md 3.5/6.1` 与 E175/E176，但该文档不在仓库内（全仓检索无此文件），评审者无法核对“除以 cp 次”的推导与适用边界。建议把关键推导与适用条件直接写入代码注释，或随 PR 一并提交该文档。

**Comment by risemeup1111** [comment]
### Paddle-Bot Review Board （review完成)

| 序号 | 位置 | 优先级 | 规则来源 | 状态 |
| --- | --- | --- | --- | --- |
| 1 | <a href="https://github.com/PaddlePaddle/PaddleFleet/pull/2083#discussion_r4034023311">CP scatter 反向的 per&#45;token 梯度除法分支无调用点</a> | P1 | <a href="https://github.com/ShigureLab/nyanpasu/blob/74655d0758cd9473ef9d07c7be8669538d002f0c/packages/nyanpasu-github-reviewer/src/nyanpasu_github_reviewer/rules/review.md">默认规则</a> | 🚧 |
| 2 | <a href="https://github.com/PaddlePaddle/PaddleFleet/pull/2083#discussion_r4034026845">per&#45;token loss 只做原始求和，缺少全局 token 分母</a> | P1 | <a href="https://github.com/ShigureLab/nyanpasu/blob/74655d0758cd9473ef9d07c7be8669538d002f0c/packages/nyanpasu-github-reviewer/src/nyanpasu_github_reviewer/rules/review.md">默认规则</a> | 🚧 |
| 3 | <a href="https://github.com/PaddlePaddle/PaddleFleet/pull/2083#discussion_r4034027881">改动未通过 ruff 检查，Codestyle Check 失败</a> | P2 | <a href="https://github.com/ShigureLab/nyanpasu/blob/74655d0758cd9473ef9d07c7be8669538d002f0c/packages/nyanpasu-github-reviewer/src/nyanpasu_github_reviewer/rules/review.md">默认规则</a> | ✅ |
| 4 | <a href="https://github.com/PaddlePaddle/PaddleFleet/pull/2083#discussion_r4034029173">calculate&#95;per&#95;token&#95;loss 开关缺少测试覆盖</a> | P2 | <a href="https://github.com/ShigureLab/nyanpasu/blob/74655d0758cd9473ef9d07c7be8669538d002f0c/packages/nyanpasu-github-reviewer/src/nyanpasu_github_reviewer/rules/review.md">默认规则</a> | 🚧 |
| 5 | <a href="https://github.com/PaddlePaddle/PaddleFleet/pull/2083#discussion_r4034030230">注释引用的 loss&#95;grad&#95;normalization&#46;md 不在仓库内</a> | P3 | <a href="https://github.com/ShigureLab/nyanpasu/blob/74655d0758cd9473ef9d07c7be8669538d002f0c/packages/nyanpasu-github-reviewer/src/nyanpasu_github_reviewer/rules/review.md">默认规则</a> | 🚧 |


<sub>Powered by Nyanpasu claude with Opus 4&#46;8 默认推理级别&#44; please check the suggestions carefully&#46;</sub>

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
<div align="right">
   <sup>Powered by Nyanpasu with DeepSeek-V4.1-Flash high, please check the suggestions carefully.</sup>
</div>

File: src/paddlefleet/context_parallel_utils.py (hunk 1 above)
Comment (risemeup1111) [reply src/paddlefleet/context_parallel_utils.py]: 当前 head 8afe851b 仍未解决：新增提交只在 language_loss.py 增加了 print-only 的展示统计（detached，不影响梯度），context_parallel_utils.py 未改动。`set_per_token_grad_scale()` 依旧没有任何调用点，`_PER_TOKEN_GRAD_SCALE` 恒为 `False`，本处 `grad / nranks` 分支仍不可达。问题维持 P1 未解决。

File: src/paddlefleet/models/common/language_loss/language_loss.py (hunk 2 above)
Comment (risemeup1111) [reply src/paddlefleet/models/common/language_loss/language_loss.py:554]: 当前 head 8afe851b 仍未解决：新增提交补充的是 print-only 的展示统计（`pop_per_token_display` 等，detached，仅用于日志打印），并未补上梯度/loss 的全局 token 分母消费逻辑——`pop_per_token_local_count()` 仍无调用点，`_PER_TOKEN_LOCAL_COUNT` 只累加不读取。开关打开后本 rank 分片原始求和导致的 loss/梯度尺度问题不变。问题维持 P1 未解决。

File: src/paddlefleet/context_parallel_utils.py (hunk 3 above)
Comment (risemeup1111) [reply src/paddlefleet/context_parallel_utils.py]: 当前 head 8afe851b 仍未通过：新增提交虽把 language_loss.py 的一处调用折为单行，但 `ruff format` 仍会重排 context_parallel_utils.py、language_loss.py、gpt_embedding.py 三个文件，Codestyle Check 在当前 head 仍为 FAILURE。请本地执行 `pre-commit run --all-files`（line-length=80）后重新推送。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

File: src/paddlefleet/context_parallel_utils.py (hunk 1 above)
Comment (Wennie396) [reply src/paddlefleet/context_parallel_utils.py]: set_per_token_grad_scale这个会在trainer端调用，当前trainer并不在本仓库

File: src/paddlefleet/models/common/language_loss/language_loss.py (hunk 2 above)
Comment (Wennie396) [reply src/paddlefleet/models/common/language_loss/language_loss.py:554]: 这些接口也是在trainer端调用，trainer不在本仓库

File: src/paddlefleet/context_parallel_utils.py (hunk 1 above)
Comment (risemeup1111) [reply src/paddlefleet/context_parallel_utils.py]: 感谢说明。理解 `set_per_token_grad_scale()` 由 trainer 侧调用、而 trainer 不在本仓库，这与本条最初建议的「由外部调用方负责」一致。仅有一点残留风险：本仓库内 `_PER_TOKEN_GRAD_SCALE` 恒为 `False`、`grad / nranks` 分支不可达且无用例覆盖，正确性完全依赖外部 trainer 的接线——若开启 `calculate_per_token_loss` 却未在 trainer 侧调用该 setter，embedding 梯度会被静默放大 cp 倍且无任何告警。建议至少在本仓库内把这一跨仓契约写入 setter 的 docstring 或特性文档（注明调用方与前置条件）。因无法在本仓库内验证，本条暂维持 P1 未解决。

File: src/paddlefleet/models/common/language_loss/language_loss.py (hunk 2 above)
Comment (risemeup1111) [reply src/paddlefleet/models/common/language_loss/language_loss.py:554]: 感谢说明。理解 `pop_per_token_local_count()` 等消费接口由 trainer 侧调用、而 trainer 不在本仓库。残留风险同前一条：本仓库内这些计数只累加、无任何读取/归一化调用点，开关打开后本 rank 分片的原始 token 求和会直接改变 loss/梯度尺度，正确性完全依赖外部 trainer 的归一化接线——若 trainer 未配套调用，用户仅打开 `calculate_per_token_loss` 就会得到尺度偏差且无告警。建议在本仓库内把该跨仓契约写入相关函数 docstring 或特性文档（注明调用方、调用时机与前置条件）。因无法在本仓库内验证，本条暂维持 P1 未解决。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

File: src/paddlefleet/models/common/language_loss/language_loss.py (hunk 2 above)
Comment (risemeup1111) [reply src/paddlefleet/models/common/language_loss/language_loss.py:554]: 新提交 0522a3ed 已在本仓库内补齐较完整的 per-token 归一化配套：`_per_token_count_enabled` / `suppress_per_token_count` 使 T_global 只统计主 lm 头（排除 MTP 头与 recompute 重放），MTP 头经 `renorm_mtp_head` 按各自 token 数回缩，aux/z-loss 也提升到 sum 尺度，并在 `__init__` / `forward_impl` 对不支持的组合（train_mtp_only、mtp_distillation_loss、eval 等）加了构造期/前向断言，明显降低了「打开即静默出错」的风险，感谢。仍有两点未闭环：(1) 最终 `1/T_global` 除法与 `pop_per_token_local_count()` 的消费仍在仓外 trainer，本仓库内无法端到端验证正确性；(2) 该开关路径仍无任何测试覆盖（见另一条 P2）。故本条暂维持 P1 未解决，待 trainer 侧配套合入或补充可在本仓验证的用例后再关闭。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

File: src/paddlefleet/context_parallel_utils.py (hunk 3 above)
Comment (risemeup1111) [reply src/paddlefleet/context_parallel_utils.py]: 新提交 bb16e00b（fix codestyle）已修复：当前 head 下 `ruff format --check` 与 `ruff check` 对本 PR 改动的全部文件均通过，Codestyle 格式问题已解决。本条关闭。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
